File: ros2/src/talker_listener/src/listener.cpp
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

class Listener : public rclcpp::Node {
private:
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_;

  void subscribe_message(const std_msgs::msg::String::SharedPtr msg) {
    RCLCPP_INFO(this->get_logger(), "Received: %s", msg->data.c_str());
  }

public:
  Listener() : Node("listener") {
    subscriber_ = this->create_subscription<std_msgs::msg::String>(
        "/chatter", 10,
        std::bind(&Listener::subscribe_message, this, std::placeholders::_1));
  };
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
  return 0;
}

File: ros2/src/talker_listener_py/talker_listener_py/listener.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class Listener(Node):
    def __init__(self):
        super().__init__('listener')
        self.subscriber_ = self.create_subscription(String, "/chatter", self.subscribe_message, 10)

    def subscribe_message(self, msg: String) -> None:
        self.get_logger().info(f"Received: {msg.data}")

def main() -> None: 
    rclpy.init()
    rclpy.spin(Listener())
    rclpy.shutdown()

if __name__ == "__main__":
    main()
